src/ast/mod.rs

```diff
@@ -745,10 +745,10 @@ pub enum Expr {
     /// ```
     /// [1]: https://duckdb.org/docs/sql/data_types/struct#creating-structs
     Dictionary(Vec<DictionaryField>),
-    /// An array index expression e.g. `(ARRAY[1, 2])[1]` or `(current_schemas(FALSE))[1]`
-    ArrayIndex {
-        obj: Box<Expr>,
-        indexes: Vec<Expr>,
+    /// An access of nested data using subscript syntax, for example `array[2]`.
+    Subscript {
+        expr: Box<Expr>,
+        subscript: Box<Subscript>,
     },
```

Comment:
For the last comment, datafusion has a method:

```rust
    fn plan_indices(
        &self,
        expr: SQLExpr,
        schema: &DFSchema,
        planner_context: &mut PlannerContext,
    ) -> Result<GetFieldAccess> {
```

where takes only the `subscript` part. And:

```rust
    fn plan_indexed(
        &self,
        expr: Expr,
        mut keys: Vec<SQLExpr>,
        schema: &DFSchema,
        planner_context: &mut PlannerContext,
    ) -> Result<Expr> {
```

where the `keys` is the `subscript` part (`vec![subscript]`). The logic is shared with MapAccess so I don't think it's easy to refactor.

Reply:
We may still have:

```rust
    ArrayIndex {
        obj: Box<Expr>,
        indexes: Vec<Expr>,
    },
```

which a new variant:

```
Subscript { ... }
```

that becomes the `indexes` part.

Reply:
I have an idea how to fix this -- working on it